d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -242,8 +220,50 @@ void execute()
         return;
       }
 
-      LOG.info("{}/{} Starting to warm up service {}", new Object[]{_requestStartedCount.incrementAndGet(), _serviceNames.size(), serviceName});
+      DualReadModeProvider.DualReadMode dualReadMode = DualReadModeProvider.DualReadMode.OLD_LB_ONLY;
+
+      if (_dualReadStateManager != null)
+      {
+        DualReadModeProvider dualReadModeProvider = _dualReadStateManager.getDualReadModeProvider();
+        dualReadMode = dualReadModeProvider.getDualReadMode(serviceName);
+        _dualReadStateManager.updateService(serviceName, dualReadMode);
+
+        // dual read mode can be fine-tuned per service, check if the specific mode for this service should start
+        // the warm-up. If not, skip this service.
+        if (!isModeToWarmUp(dualReadMode, _isIndis))
+        {
+          return;
```

Comment:
It shouldn't be possible to return from here since this would kill one of the warmup chains ("concurrent requests"). Eg. if there are 10 services per batch, one of them would get killed right here and the next batch would have only 9, since the 10th one was supposed to be executed following this one's callback completion.